include/tinycompress/compress_ops.h

```diff
@@ -0,0 +1,40 @@
+/* SPDX-License-Identifier: (LGPL-2.1-only OR BSD-3-Clause) */
+/* Copyright (c) 2020 The Linux Foundation. All rights reserved. */
+
+#ifndef __COMPRESS_OPS_H__
+#define __COMPRESS_OPS_H__
+
+#include "sound/compress_params.h"
+#include "sound/compress_offload.h"
+#include "tinycompress.h"
+
+struct compress_ops {
+	void *(*open)(unsigned int card, unsigned int device,
+			unsigned int flags, struct compr_config *config);
+	void (*close)(void *compress_data);
+	int (*get_hpointer)(void *compress_data,
+			unsigned int *avail, struct timespec *tstamp);
+	int (*get_tstamp)(void *compress_data,
+			unsigned int *samples, unsigned int *sampling_rate);
+	int (*write)(void *compress_data, const void *buf, size_t size);
+	int (*read)(void *compress_data, void *buf, size_t size);
+	int (*start)(void *compress_data);
+	int (*stop)(void *compress_data);
+	int (*pause)(void *compress_data);
+	int (*resume)(void *compress_data);
+	int (*drain)(void *compress_data);
+	int (*partial_drain)(void *compress_data);
+	int (*next_track)(void *compress_data);
+	int (*set_gapless_metadata)(void *compress_data,
+			struct compr_gapless_mdata *mdata);
+	void (*set_max_poll_wait)(void *compress_data, int milliseconds);
+	void (*set_nonblock)(void *compress_data, int nonblock);
+	int (*wait)(void *compress_data, int timeout_ms);
```

Comment:
hmm what is the purpose of wait() here?

Reply:
Same - https://github.com/alsa-project/tinycompress/blob/master/include/tinycompress/tinycompress.h#L270
```
/* Wait for ring buffer to ready for next read or write */
int compress_wait(struct compress *compress, int timeout_ms);
```

Reply:
But do we want plugins to wait, why cant this be handled in lib... Right now there is 1:1 mapping bw plugin APIs and compress APIs, is this right way to go, we should probably do a bit more here and let plugins be lighter... What do you think... @perexg how do you recommend this kind of split? How does alsa-lib do it?

Reply:
Initial comment from @perexg was to abstract the APIs at upper layer and do not implement kernel APIs like ioctl/poll. We earlier had ops->poll() implemented in compress_wait() - https://github.com/alsa-project/tinycompress/commit/02f68a0340b802ff1a2be7a93265f79e820c2d3e#diff-2441a41ceb86f153d1dd7b2702b49d7cR614

Reply:
Quite right and I am questioning that if the plugin interface sholud follow compress ops implemented, should it not be at a higher level with things like poll/wait/nonblock handled here...? I would think that the plugin may not care much about these things... and we should have open/read/write/pause/resume kind of higher level functions with plugins. what do you think?

Reply:
nonblock is used in compress_write() and struct compress is not shared with plugin library. So, having it in compress.c will not work unless we have common code handling done for compress_write() too.
Regarding wait(), it actually does poll on compress fd for HW node and poll is also used during compress_read/write(). If we want to have common handling only for wait() (by probably having an API get_poll_fd()), then plugin library will be restricted to create some fd (eventfd/pipefd/etc) and use it to unblock the poll. This will put limitation in plugin implementation to unblock compress_wait() only using fd.
Please suggest.

Reply:
> Quite right and I am questioning that if the plugin interface sholud follow compress ops implemented, should it not be at a higher level with things like poll/wait/nonblock handled here...? I would think that the plugin may not care much about these things... and we should have open/read/write/pause/resume kind of higher level functions with plugins. what do you think?

The exported API is simple and we should not expect that there's one file descriptor on the bottom. In this case, it's better to implement (overwrite) everything in the plugin.